packages/devtools_app/lib/src/memory/flutter/memory_chart.dart

```diff
@@ -0,0 +1,332 @@
+        highLightPerTapEnabled: true,
+        backgroundColor: ColorUtils.WHITE,
+        drawGridBackground: false,
+        dragXEnabled: true,
+        dragYEnabled: true,
+        scaleXEnabled: true,
+        scaleYEnabled: true,
+        pinchZoomEnabled: false,
+        description: desc);
+
+    // Compute padding around chart.
+    _chartController.setViewPortOffsets(50, 10, 10, 30);
+  }
+
+  final List<Entry> _used = <Entry>[];
+  final List<Entry> _capacity = <Entry>[];
+  final List<Entry> _externalHeap = <Entry>[];
+
+  var _img;
+
+  LineDataSet usedHeapSet;
+  LineDataSet capacityHeapSet;
+  LineDataSet externalMemorySet;
+
+  _loadImage() async => ImageLoader.loadImage('assets/img/star.png');
+
+  int timerDataIndex = 0;
+  Timer _timer;
+  Future<void> startTimer(
+    List<Entry> capacity,
+    List<Entry> used,
+    List<Entry> externalHeap,
+  ) async {
+    if (_img == null) await _loadImage();
+
+    used.clear();
+    capacity.clear();
+    externalHeap.clear();
+
+    // Prime the data.
+    if (timerDataIndex < externalMemoryData.length) {
+      int x;
+      int y;
+
+      x = externalMemoryData[timerDataIndex];
+      y = externalMemoryData[timerDataIndex + 1];
+      externalHeap.add(Entry(x: x.toDouble(), y: y.toDouble(), icon: _img));
+
+      x = usedHeapData[timerDataIndex];
+      y = usedHeapData[timerDataIndex + 1] +
+          externalMemoryData[timerDataIndex + 1];
+      used.add(Entry(x: x.toDouble(), y: y.toDouble(), icon: _img));
+
+      x = heapCapacityData[timerDataIndex];
+      y = heapCapacityData[timerDataIndex + 1];
+      capacity.add(Entry(x: x.toDouble(), y: y.toDouble(), icon: _img));
+
+      timerDataIndex += 2;
+    }
+
+    // Average rate is ~500-600 ms?
+    _timer = Timer.periodic(const Duration(milliseconds: 200), (Timer timer) {
+      // Pause pressed stop pumping out data simulating a live feed.
+      if (widget.memoryController.paused) return;
+
+      if (timerDataIndex < externalMemoryData.length) {
+        int x;
+        int y;
+
+        x = externalMemoryData[timerDataIndex];
+        y = externalMemoryData[timerDataIndex + 1];
+        externalHeap.add(Entry(x: x.toDouble(), y: y.toDouble(), icon: _img));
+
+        x = usedHeapData[timerDataIndex];
+        y = usedHeapData[timerDataIndex + 1] +
+            externalMemoryData[timerDataIndex + 1];
+        used.add(Entry(x: x.toDouble(), y: y.toDouble(), icon: _img));
+
+        x = heapCapacityData[timerDataIndex];
+        y = heapCapacityData[timerDataIndex + 1];
+        capacity.add(Entry(x: x.toDouble(), y: y.toDouble(), icon: _img));
+
+        timerDataIndex += 2;
+
+        // Signal data has changed.
+        usedHeapSet.notifyDataSetChanged();
+        capacityHeapSet.notifyDataSetChanged();
+        externalMemorySet.notifyDataSetChanged();
+
+        _chartController.data = LineData.fromList(
+            []..add(usedHeapSet)..add(externalMemorySet)..add(capacityHeapSet));
+
+        setState(() {});
```

Comment:
There are cases where an empty setState is OK, but they are rare.  The changes to the fields on this class should be guarded by setState calls anywhere except for the state lifecycle methods (initState, didUpdateWidget, dispose, etc)

Reply:
I did fix that.  However, I did still use it in one place where I simulate the live feed I thought it was simpler to made a bunch of chart updates then after those changes are made call setState with an empty function.  Could have peppered with setState or wrap a whole bunch of code with setState.